### ISSUE TITLE
Can't package pyusb, 'Unable to find "libusb-1.0.so.0"'

### ISSUE THREAD

**Issue by joelhoisko** [pyinstaller/pyinstaller#2633]
To reproduce

1. `virtualenv env`
`source env/bin/activate`
`pip install https://github.com/pyinstaller/pyinstaller/archive/develop.zip`
`pip install pyusb`
`pip freeze > requirements.txt`
2. Create simple python file which imports pyusb
test.py
```
import usb.core
print 'Success!
```
3. Run the python file normally
```
$ python test.py
Success!
```
4. Package it with pyinstaller
```
pyinstaller --log-level=DEBUG test.py &> log.txt
```
log.txt
```
14 INFO: PyInstaller: 3.3.dev0+5fd013bfe
14 INFO: Python: 2.7.12+
14 INFO: Platform: Linux-4.8.0-53-generic-x86_64-with-Ubuntu-16.10-yakkety
14 INFO: wrote /home/joel/Code/random/pyinstaller/test.spec
15 DEBUG: Testing for UPX ...
17 INFO: UPX is not available.
17 DEBUG: script: /home/joel/Code/random/pyinstaller/test.py
17 INFO: Extending PYTHONPATH with paths
['/home/joel/Code/random/pyinstaller', '/home/joel/Code/random/pyinstaller']
17 INFO: checking Analysis
17 INFO: Building Analysis because out00-Analysis.toc is non existent
17 INFO: Initializing module dependency graph...
18 INFO: Initializing module graph hooks...
19 DEBUG: Hidden import: codecs
65 DEBUG: Hidden import 'codecs' already found
65 INFO: running Analysis out00-Analysis.toc
65 DEBUG: Analyzing /home/joel/Code/random/pyinstaller/env/bin/python
71 DEBUG: Skipping libdl.so.2 dependency of python
71 DEBUG: Skipping libc.so.6 dependency of python
71 DEBUG: Skipping libutil.so.1 dependency of python
71 DEBUG: Skipping libpthread.so.0 dependency of python
71 DEBUG: Skipping libm.so.6 dependency of python
72 DEBUG: Adding libz.so.1 dependency of python from /lib/x86_64-linux-gnu/libz.so.1
72 DEBUG: Analyzing /lib/x86_64-linux-gnu/libz.so.1
78 DEBUG: Skipping libc.so.6 dependency of libz.so.1
78 INFO: Caching module hooks...
81 INFO: Analyzing /home/joel/Code/random/pyinstaller/test.py
1078 INFO: Loading module hooks...
1079 INFO: Loading module hook "hook-usb.py"...
Unable to find "libusb-1.0.so.0" when adding binary and data files.
```
This causes the packaging process to halt with an empty `dist/` and `build/test/` folders.

I dug around and found out, that `hook-usb.py` reports the binaries to be `[('libusb-1.0.so.0', '')]` and I think this causes that  `src_root_path_or_glob`(`utils.py`, lines 450+) to just contain 'libusb1.0.so.0', which produces just `[]`after it get's run through the `glob.glob()` and not the real path, which I think is`/lib/x86_64-linux-gnu/libusb-1.0.so.0` on my computer.

Environment:
```
Python 2.7.12+
Ubuntu 16.10
Virtualenv 15.1.0
```
requirements.txt
```
appdirs==1.4.3
packaging==16.8
PyInstaller==3.3.dev0+5fd013bfe
pyparsing==2.2.0
pyusb==1.0.0
six==1.10.0
```

Also, this seems to be somewhat related to #1682 and commit https://github.com/pyinstaller/pyinstaller/commit/23901eb1ce7e60f1aac2a73e6cb3b02f1a34b9b1

**Comment by 494895913** [comment]
I also have this problem

Can you tell me how to solve it?

**Comment by joelhoisko** [comment]
I haven't tried to solve it/dig deeper into yet, as my project doesn't yet depend on `pyusb`.

**Comment by htgoebel** [comment]
Please help tracking this down:
- Add `import pdb; pdb.set_trace()` near the top of `hook-usb.py` (just behind the import statements)
- In the shell run `export PYUSB_DEBUG=debug`
- run PyInstaller `pyinstaller test.py` (no output-redirection)
- step through `hook-usb.py` to gain more insights.

Thanks.

**Comment by joelhoisko** [comment]
Sorry for the delay, but I managed to find something to get me forward.

The custom mechanism of `hook-usb.py` as defined in [here](https://github.com/pyinstaller/pyinstaller/commit/23901eb1ce7e60f1aac2a73e6cb3b02f1a34b9b1#diff-8e99511b3f6d0c631ac915fd6ca91a80R50) seems to find the right binaries for my system. The 'pyusb library locator' (lines 29-46) makes the `binaries` to be `[('libusb-1.0.so.0', '')]`, meaning the branch `if not binaries:` on line 50 isn't executed even though we don't have a path to the binaries.

I changed line 50 to `if binaries:` for testing purposes and `$ pyinstaller test.py` runs fine and the usb module works as intended. I also tested it on a small script (writes to an ESC/POS printer with pyusb) and it worked fine.

So yeah, I don't know the underlying issue of why the `usblib._name` doesn't resolve into the full path (I guess it should?) but this 'hack' circumvents the problem.

**Comment by joelhoisko** [comment]
The fix by themadinventor does indeed seem to work for me.

**Comment by jxltom** [comment]
In Linux #5de8a59 works fine. In Windows 7 Enterprise, assert len(binaries[0]) == 3 fails when using Conda's python35 env with libusb...

**Comment by htgoebel** [comment]
@jxltom Please open an issue providing all information required for analysis, esp. the traceback.

**Comment by charlesnicholson** [comment]
We hit this issue as well. Here's a repository you can clone to see the failure:

Python 3.6, Ubuntu 16.04:
https://github.com/charlesnicholson/pyinstaller-linux-pyusb-bug-repro

As reported above, the discovered backend isn't anchored in any path, and `_resolveCtypesImports` is never called on it to compute the absolute path.

Note also that on Ubuntu 14.04 we see PyInstaller crashes coming from usb-hook.py:

usb-hook.py first attempts to get the backend by importing and using `usb.backend`:
https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/hooks/hook-usb.py#L31

usb gets unloaded when the module is finalized, and `libusb_exit` causes a segmentation fault inside of PyInstaller, which then crashes.

If detecting the usb shared library by way of the pyusb backend is required, it would be nice to be able to configure the hook to not attempt this method, and instead use the `ctypes.util.find_library` approach.

**Comment by charlesnicholson** [comment]
BTW I'm happy to commit time to writing a patch if there's a maintainer-approved solution.

**Comment by charlesnicholson** [comment]
I've added some print instrumentation to my repro repository that shows the exact problem. Check out the logs at the bottom of the README.md at
https://github.com/charlesnicholson/pyinstaller-linux-pyusb-bug-repro

The PyInstaller fork that I've instrumented is here:
https://github.com/charlesnicholson/pyinstaller/tree/usb-hook-linux-repro

The instrumented hook file is here:
https://github.com/charlesnicholson/pyinstaller/commit/f4297eb61c0d6e3669afb74432a4c316ba7adbe5

I'm curious what the motivation is for the first half of the file that engages pyusb? It seems like if your libusb is somewhere that only pyusb can find and not in your LD_LIBRARY_PATH etc, maybe that's worth a custom entry in your spec file?

It seems like the entire hook might be simpler and just as full-featured if the bottom half of the file that uses `ctypes.util.find_library` was the only approach used?

In my hacked-up fork that unblocked my project, I simply wrapped the top half of the file in "if we're not on Linux" logic and it seems to work fine. That's probably not a great long-term solution though :)

**Comment by htgoebel** [comment]
> I'm happy to commit time to writing a patch if there's a maintainer-approved solution.

Great! Please provide a pull-request, so we can easily see, review and comment your changes. Thanks.

**Comment by charlesnicholson** [comment]
Sure, happy to. My only initial question is that previously in this thread, this solution was posted:
https://github.com/jxltom/pyinstaller/commit/8789584dda3c19ebcb6819307e4d749c905931fa

Which (correctly) calls `_resolveCtypesImports` on the basename of the discovered library.

I think that's the correct fix for the case where you /want/ to use the pyusb backend.

My open question is, what's the advantage of engaging the pyusb backend? It seems like it should always give you the same results that `ctypes.util.find_library` does, but is more heavyweight in that it actively loads various libusb implementations.

I'm happy to open a new PR that simply deletes the top half of the file, but I think that might be bulldozing past some use cases.

I think the least intrusive PR would be a combination of what's listed above, plus something like a check for an environment variable like `PYINSTALLER_HOOK_USB_DONT_SEARCH_WITH_PYUSB` so that the hook's default behavior doesn't change, but can be overridden.

Is this a reasonable approach?

**Comment by charlesnicholson** [comment]
Opened https://github.com/pyinstaller/pyinstaller/pull/3269 with fixes, feedback welcome!

**Comment by charlesnicholson** [comment]
I'm seeing some [build failures](https://travis-ci.org/pyinstaller/pyinstaller/builds/337087283) that don't look like they're related to my work. Python 2.7 Linux, 3.4 Linux, 3.6 Mac, and 3.6 Linux all pass.

Maybe a core team member or maintainer could help me understand if these are related, and how to move forward?

Python 3.3 Linux:
```
$ py.test -n 3 --maxfail 3 tests/functional/test_libraries.py tests/functional/test_hooks
usage: py.test [options] [file_or_dir] [file_or_dir] [...]
py.test: error: unrecognized arguments: -n tests/functional/test_libraries.py tests/functional/test_hooks
  inifile: None
  rootdir: /home/travis/build/pyinstaller/pyinstaller
The command "py.test -n 3 --maxfail 3 tests/functional/test_libraries.py tests/functional/test_hooks" exited with 2.
```

Python 3.5 Linux:
```
=================================== FAILURES ===================================
________________________ test_pandas_extension[onefile] ________________________
[gw0] linux -- Python 3.5.4 /home/travis/virtualenv/python3.5.4/bin/python
pyi_builder = <conftest.AppBuilder object at 0x7fd4643e2ef0>
    @importorskip('pandas')
    def test_pandas_extension(pyi_builder):
        # Tests that C extension 'pandas.lib' is properly bundled. Issue #1580.
        pyi_builder.test_source(
            """
            from pandas.lib import is_float
            assert is_float(1) == 0
>           """)
E       Failed: Running exe /tmp/pytest-of-travis/pytest-0/popen-gw0/test_pandas_extension_onefile_0/dist/test_pandas_extension failed with return-code 1.
/home/travis/build/pyinstaller/pyinstaller/tests/functional/test_libraries.py:806: Failed
----------------------------- Captured stdout call -----------------------------
------- Starting build. -------
Syntax error in  /home/travis/virtualenv/python3.5.4/lib/python3.5/site-packages/jinja2/asyncsupport.py
("'yield' inside async function", ('/home/travis/virtualenv/python3.5.4/lib/python3.5/site-packages/jinja2/asyncsupport.py', 35, 12, '            yield event\n'))
------- Build finshed, now running executable. -------
RUNNING:  '/tmp/pytest-of-travis/pytest-0/popen-gw0/test_pandas_extension_onefile_0/dist/test_pandas_extension' , args:  ['./test_pandas_extension']
TIMED OUT:  '/tmp/pytest-of-travis/pytest-0/popen-gw0/test_pandas_extension_onefile_0/dist/test_pandas_extension' , args:  ['./test_pandas_extension']
----------------------------- Captured stderr call -----------------------------
```

**Comment by mofahead** [comment]
I just got hit with this problem too. Awesome that it's being worked on!

**Comment by charlesnicholson** [comment]
@htgoebel It's great that the build system is better and that the first fix went in!

Please also take a look at my work in https://github.com/pyinstaller/pyinstaller/pull/3269. It cleans up duplicated logic, simplifies the usb tests, and allows for the ubuntu 14.04 usb hook crash to be avoided by an environment variable.

It's a more thorough set of fixes and improvements to the usb hook module than what you just merged.
